lsp-serenata.el

```diff
@@ -39,21 +39,56 @@ It can be downloaded from https://gitlab.com/Serenata/Serenata/-/releases."
   :group 'lsp-serenata
   :type 'file)
 
+(defcustom lsp-serenata-php-version
+  7.3
+  "Allows you to specify the PHP version your project is written in.
+At the moment this directive is still ignored, but it will
+influence functionality such as refactoring in the future, where
+older PHP versions may not support scalar type hints, which may
+then be omitted from places such as getters and setters."
+  :group 'lsp-serenata
+  :type 'number)
+
+(defcustom lsp-serenata-file-extensions
+  ["php"]
+  "List of file extensions (without dot) to process.
+Files that do not match this whitelist will be ignored during
+indexing.  Usually you'll want to set this to at least include
+php, as it is the most common PHP extension.  phpt is not
+included by default as it is often used to contain test code that
+is not directly part of the code.  Note that for existing
+projects, removing extensions will not not automatically prune
+files having them from the index if they are already present.
+Adding new ones will cause the files having them to be picked up
+on the next project initialization."
+  :group 'lsp-serenata
+  :type 'array)
+
 (defun lsp-serenata-server-start-fun (port)
   "Define serenata start function, it requires a PORT."
   `(,lsp-serenata-server-path
     "-u" ,(number-to-string port)))
 
+(defun lsp-serenata-init-options ()
+  "Init options for lsp-serenata."
+  `(
+    :phpVersion ,lsp-serenata-php-version
+    :fileExtensions ,lsp-serenata-file-extensions))
+
 (lsp-register-client
  (make-lsp-client
   :new-connection (lsp-tcp-connection 'lsp-serenata-server-start-fun)
   :major-modes '(php-mode)
   :priority -2
-  :server-id 'serenata
+  :notification-handlers (ht ("serenata/didProgressIndexing"
+			      (lambda (server data)
+				(message "%s" (ht-get data "info")))))
```

Comment:
We currently support ht and plists, so I think it's better to define an interface for this notification like [here](https://github.com/emacs-lsp/lsp-mode/blob/master/lsp-protocol.el#L261) and use `(lsp-my-interface-info data)` here.

Reply:
Sorry, I dont understand the correct way to do it, 

(lsp-interface (serenata:ProcessStartNotification nil (:info))

  :notification-handlers 'lsp:serenata-process-start-notification-info 

That code give me some error.

I think I'm doing something wrong.

Reply:
Basically, we created a way to define some protocol for elisp of the LSP protocol and other server specific protocols.
If you have a Notification that returns an object of `ProcessStartNotification` that contains some fields like `info` as required and `range` as optional, for example, you can define an interface like:
```elisp
(lsp-interface (serenata:ProcessStartNotification (:info) (:range)))
```
In your case maybe: 
```elisp
(lsp-interface (serenata:ProcessStartNotification (:info) nil))
```

After that, when you require `lsp-protocol` in the file you can use `(lsp:serenata-process-start-notification-info data)`

Reply:
and one more thing: the `ht` after notification handlers stays as it is. The only thing that is changed is the way you access the structures coming from the language server. You may take a look at the unit tests for example usages: https://github.com/emacs-lsp/lsp-mode/blob/master/test/lsp-protocol-test.el

Reply:
Ok ,I think I figure out how to solve the thing :+1: 

Thanks for the tips!